### ISSUE TITLE
Virtual text to the right of the line that fits exactly on screen causes that line and all lines below it not to be displayed

### ISSUE THREAD

**Issue by mgedmin** [vim/vim#12213]
### Steps to reproduce

When vim-ale puts a bit of virtual text after the end of the line, and the length of the actual text exactly matches the window width, Vim gets confused and hides the entire line (and usually the lines below it):

![paveikslas](https://user-images.githubusercontent.com/159967/229134019-e5c3c7b9-0549-4125-8d93-ff244c0d5e0d.png)

This requires `:set nowrap` and is a bit easier to see with `:set display=`.

I've managed to reproduce this or something very like this (I don't see the `@` characters, but the text itself disappears):

1. Create a `repro.vim` with
  
```
40vsplit
0put ='here is some text more text even *more*:'
put ='  more text!'
put ='  more text!'
put ='  more text!'
put ='  more text!'
set nomodified " make :qa work for convenience
set nowrap
set display=
set list listchars=extends:>
call prop_type_add("bug", {"highlight": "Error"})
call prop_add(1, 0, {"type": "bug", "text": "# virtual text that is longer than the actual text", "text_padding_left": 1})
```

2. Run `vim --clean -S repro.vim`

Observe:

![paveikslas](https://user-images.githubusercontent.com/159967/229135560-cc145ac7-4e5e-4812-885d-f750a3820667.png)

### Expected behaviour

This is what things look like when the window is one character wider than the text:

![paveikslas](https://user-images.githubusercontent.com/159967/229134554-f07f088e-4a6d-44c7-8618-7792b7354ca6.png)

and in my synthetic example:

![paveikslas](https://user-images.githubusercontent.com/159967/229135688-48f268ee-1776-4e19-baaf-27192ea47fe9.png)

### Version of Vim

9.0.1392

### Environment

OS: Ubuntu 22.10
Terminal: gnome-terminal 3.46.2 using VTE 0.70.0
$TERM: xterm-256color
Shell: bash

**Comment by mgedmin** [comment]
Forgot to mention one possibly related detail: the `listchars` `extends` character (`>`) is not drawn when the window width exactly matches text width, despite there being virtual text after it.  If I increase the size of the window, the `>` re-appears.  I'm not sure if I should be filing a separate bug for it, or if it's caused by the same underlying drawing bug.

**Comment by bfrg** [comment]
Just wanted to mention that the error remains with `nolist`:

```vim
vim9script
# 0123456789 0123456789 0123456789 0123456789

:45vsplit
set nowrap
prop_type_add('test', {highlight: 'Error'})
prop_add(2, 0, {
    text_align: 'after',
    text: 'The quick brown fox jumps over the lazy dog',
    type: 'test'
})
```

1. Run `vim --clean test.vim -S test.vim`.
2. Result: The content in the left split is not displayed correctly.

Screenshot:

![screenshot-2023-03-31_204036](https://user-images.githubusercontent.com/6266600/229203617-7df38c68-1c70-4c12-b004-bf203580a49b.png)


After pressing `CTRL-W >` (or `CTRL-W <`) everything is as expected:

![screenshot-2023-03-31_204043](https://user-images.githubusercontent.com/6266600/229203618-dc2a0524-24eb-49c3-8ab5-8ea5f8153bf9.png)

**Comment by dylanahsmith** [comment]
> Forgot to mention one possibly related detail: the `listchars` `extends` character (`>`) is not drawn when the window width exactly matches text width, despite there being virtual text after it. If I increase the size of the window, the `>` re-appears. I'm not sure if I should be filing a separate bug for it, or if it's caused by the same underlying drawing bug.

This behaviour had already changed before I took a look at the issue, it now already shows the `extends` character on the truncated line in the provided test case, even without my proposed fix (#14307)
